src/lib/Microsoft.Health.Common/EventHub/Listeners/EventHubListener.cs

```diff
@@ -230,6 +236,129 @@ private async Task TimedFlush()
+                    {
+                        Console.WriteLine($"Current UTC: {DateTime.UtcNow}");
+                        Console.WriteLine($"Current Interval End: {_currentIntervalEnd} is greater than {DateTime.UtcNow.AddSeconds(-5)}");
+                        Console.WriteLine("Waiting...");
+                        await Task.Delay(5000);
+                    }
+
+                    // Process events within interval
+                    await ConsumeEventsInIntervalAsync();
+                    _currentIntervalStart = _currentIntervalStart.Add(_flushTimeSpan);
+                    _currentIntervalEnd = _currentIntervalEnd.Add(_flushTimeSpan);
+                    Console.WriteLine("Next Interval End: " + _currentIntervalEnd);
+                }
+            }
+
+            private DateTime GetFirstEnqueuedTime()
+            {
+                // TODO - compare across all _pendingEventData.Values?
+                foreach (var pendingData in _pendingEventData.Values)
+                {
+                    if (pendingData.Data.IsEmpty)
+                    {
+                        continue;
+                    }
+
+                    pendingData.Data.TryPeek(out var eventData);
+                    eventData.SystemProperties.TryGetValue("x-opt-enqueued-time", out var enqueuedTime);
+                    var enqueuedDateTime = Convert.ToDateTime(enqueuedTime);
+                    eventData.Dispose();
+                    return enqueuedDateTime;
+                }
+
+                // Else return min value
+                return DateTime.MinValue;
+            }
+
+            private DateTime CalculateIntervalEnd(DateTime intervalStart, TimeSpan intervalLength)
+            {
+                if (intervalStart == DateTime.MinValue)
+                {
+                    intervalStart = GetFirstEnqueuedTime();
+                }
+
+                if (intervalStart == DateTime.MinValue)
+                {
+                    return intervalStart;
+                }
+
+                var intervalEnd = intervalStart.Add(intervalLength);
+
+                return intervalEnd;
+            }
+
+            // Consume events within a certain interval
+            private async Task ConsumeEventsInIntervalAsync()
+            {
+                var intervalEnd = _currentIntervalEnd;
+
+                Console.WriteLine($"=======Flushing Events up to: {intervalEnd}=======");
+
+                // find index, then dequeue up to index.
+                foreach (var pendingData in _pendingEventData.Values)
+                {
+                    if (pendingData.Data.IsEmpty)
+                    {
+                        continue;
+                    }
+
+                    var batch = new List<EventData>();
+                    while (pendingData.Data.TryPeek(out var eventData))
+                    {
+                        eventData.SystemProperties.TryGetValue("x-opt-enqueued-time", out var enqueuedTime);

```

Comment:
You don't really want to enforce the enqueue time here.  We should be checking before we even buffer the value.

Reply:
Specifically in the ProcessEventsAsync(PartitionContext context, IEnumerable<EventData> messages) method.